### ISSUE TITLE
Weight Decay

### ISSUE THREAD

**Issue by Atticus1806** [rwth-i6/returnn_common#149]
I am trying to use weight decay but I am not sure whether i am doig it the right way, because its not added to the network dict. Below you can find a small test that fails:

```
def test_weight_decay():

  nn.reset_default_root_name_ctx()
  inputs = nn.get_extern_data(nn.Data('h', dim_tags=[nn.FeatureDim('F', 3)]))
  x = nn.Linear(nn.FeatureDim("Lin", 4))
  for param in x.parameters():
    param.weight_decay = 1.0
  forw = x(inputs)
  forw.mark_as_default_output()
  config = nn.get_returnn_config().get_complete_py_code_str(x)
  from returnn_common.tests.returnn_helpers import config_net_dict_via_serialized
  config, net_dict = config_net_dict_via_serialized(config)
  assert "L2" in net_dict["weight"]
```

**Comment by albertz** [comment]
`nn.Linear` creates the parameter lazily when you don't specify the input dim. So the parameter only exists after you called `x(inputs)` here. It's not really possible otherwise because you need to know the input dim to be able to create the param.

I just added `test_parameter_weight_decay` which does it in the correct way.

So, should we do sth about this? You are right that this might be a bit unexpected.

Maybe we could already create the param and allow to change its shape afterwards? But allowing this is tricky and could lead to other problems.

**Comment by Atticus1806** [comment]
Ah I see. While I think this is unexpected I am not sure if its worth other problems. Maybe documenting the setter method with a reference of the ordering (that it has to be done after the creation) can already do the job. Or referencing this issue or so.

**Comment by Atticus1806** [comment]
You are right.

**Comment by albertz** [comment]
I wonder a bit whether this behavior of `nn.Linear` and some other modules is actually maybe causing confusion here. Esp. this is sth a user could easily do wrong and would not notice because there is no error.

We maybe could also remove all such lazy behavior and require that all parameters are always defined in the `__init__` of a module. But this would make other code longer. PyTorch `nn.Linear` requires the input and output dim and then can create the param in `__init__`. However, later they introduced `nn.LazyLinear` because there are often situations where users would prefer the shorter variant. `layers.Dense` also behaves lazily this way.

We could also maybe check if you are iterating over such lazy non-initialized module and throw an exception then. I'm not sure if there are valid cases where you only want to iterate over some subset of the initialized parameters. Probably not. So throwing an exception then would be good because then the user will directly notice.

**Comment by Atticus1806** [comment]
>So throwing an exception then would be good because then the user will directly notice.

I agree. I think the layz behavior itself is nice and keeps the code simple. Throwing an exception is a good way to deal with it, because then the user explicitly needs to decide here (similar to broadcasting with standard ops)
